contracts/AssetPool.sol

```diff
@@ -253,13 +260,75 @@ contract AssetPool is Ownable {
         underwriterToken.burn(covAmount);
     }
 
+    /// @notice Withdraws collateral tokens to a new Asset Pool which previously
+    ///         was approved by the governance. Old underwriter tokens are burned
+    ///         in favor of new minted tokens that proof share of
+    ///         collateral tokens in a new Asset Pool.
+    function withdrawToNewAssetPool() external {
+        /* solhint-disable not-rely-on-time */
+        require(
+            address(newAssetPool) != address(0),
+            "New asset pool must be assigned"
+        );
+
+        uint256 covUnderwriterBalance = underwriterToken.balanceOf(msg.sender);
+
+        require(
+            covUnderwriterBalance > 0,
+            "Underwriter's token balance must be greater than 0"
+        );
+
+        uint256 covSupply = underwriterToken.totalSupply();
+
+        // slither-disable-next-line reentrancy-events
+        rewardsPool.withdraw();
+
+        uint256 collateralBalance = collateralToken.balanceOf(address(this));
+
+        uint256 amountToWithdraw =
+            covUnderwriterBalance.mul(collateralBalance).div(covSupply);
+
+        collateralToken.safeTransfer(address(newAssetPool), amountToWithdraw);
+        emit WithdrawalToNewAssetPoolCompleted(
+            msg.sender,
+            amountToWithdraw,
+            block.timestamp
+        );
+
+        // mappings cleanup since all the underwriter's funds are transferred to
+        // a new asset pool
+        delete withdrawalInitiatedTimestamp[msg.sender];
+        delete pendingWithdrawal[msg.sender];
```

Comment:
I am not sure about it.... If we just delete them, and there is a pending withdrawal, the underwriter will lose a part of their tokens.

Reply:
Wouldn't [this](https://github.com/keep-network/coverage-pools/blob/upgradeability-for-asset-pool/contracts/AssetPool.sol#L274) line do the trick? I take that we're interested in moving the entire balance of underwriter to a new Asset Pool. Can't we read the total balance no matter if it's pending or not and then transfer it?

Reply:
The link is pointing now to a solhint line so I am not sure what was there :) But I think we should keep upgrade and withdrawal separate for simplicity.